From 8e5bb7976b9926a8e79f5dfb1628b871fab5a66f Mon Sep 17 00:00:00 2001
From: "haotian.qu" <haotian.qu@amlogic.com>
Date: Mon, 17 Apr 2023 17:38:23 +0800
Subject: [PATCH 2/3] SWUpdate: T7C needs SWUpdate feature on kernel5.15 [1/1]

PD#SWPL-114098

Problem:
1. SWUpdate function exception, kernel 5.15 missing such as
/dev/boot,/dev/system device node.
2. The directfbrc file is the same as the file name in DirectFB.
3. T7C kernel5.4 doesn't needs swupdate

Solution:
1. Establish a device soft link.
2. Rename directfbrc file as ota_directfbrc
3. Disable swupdate feature in T7C kernel5.4

Verify:
local

Signed-off-by: haotian.qu <haotian.qu@amlogic.com>
Change-Id: Ie7d8b5ba3734e630cb4aad1d6b148338d4814a74
---
 conf/machine/mesont7c-an400-weston.config     |  2 +-
 recipes-core/images/aml-package.inc           |  1 +
 .../initramfs-recovery/init-recovery_5.15     | 13 ++++++++
 .../{directfbrc => ota_directfbrc}            |  6 ++--
 recipes-core/swupdate/aml-swupdate-ui_git.bb  | 10 +++----
 recipes-core/swupdate/swupdate/swupdate.sh    | 30 +++++++++----------
 recipes-core/swupdate/swupdate_git.bb         |  3 +-
 7 files changed, 40 insertions(+), 25 deletions(-)
 rename recipes-core/swupdate/aml-swupdate-ui/{directfbrc => ota_directfbrc} (78%)

diff --git a/conf/machine/mesont7c-an400-weston.config b/conf/machine/mesont7c-an400-weston.config
index cf8b088..4c19105 100644
--- a/conf/machine/mesont7c-an400-weston.config
+++ b/conf/machine/mesont7c-an400-weston.config
@@ -65,7 +65,7 @@ DISTRO_FEATURES_append = " aml-isp arm-isp "
 DISTRO_FEATURES_append = " amlogic-tv "
 DISTRO_FEATURES_append = " softap"
 DISTRO_FEATURES_append = " aml-libgdc"
-DISTRO_FEATURES_append = " swupdate"
+#DISTRO_FEATURES_append = " swupdate"
 DISTRO_FEATURES_append = " recovery"
 PREFERRED_VERSION_isp = "5.4"
 DISTRO_FEATURES_append = " gst-aml-imgproc"
diff --git a/recipes-core/images/aml-package.inc b/recipes-core/images/aml-package.inc
index fa30529..4bb5f3c 100644
--- a/recipes-core/images/aml-package.inc
+++ b/recipes-core/images/aml-package.inc
@@ -175,6 +175,7 @@ fi
             fi
         fi
         sed 's@boardname@${MACHINE_ARCH}@' -i ${DEPLOY_DIR_IMAGE}/sw-description
+        sed 's/_5.15//g' -i ${DEPLOY_DIR_IMAGE}/sw-description
 
         ${DEPLOY_DIR_IMAGE}/sw_package_create.sh ${DEPLOY_DIR_IMAGE}
     fi
diff --git a/recipes-core/initrdscripts/initramfs-recovery/init-recovery_5.15 b/recipes-core/initrdscripts/initramfs-recovery/init-recovery_5.15
index 1d94898..d37f835 100644
--- a/recipes-core/initrdscripts/initramfs-recovery/init-recovery_5.15
+++ b/recipes-core/initrdscripts/initramfs-recovery/init-recovery_5.15
@@ -1,6 +1,7 @@
 #!/bin/sh
 
 PATH=/sbin:/bin:/usr/sbin:/usr/bin
+MMC_DEVICE_PATH=/sys/block/mmcblk0
 
 # Copied from initramfs-framework. The core of this script probably should be
 # turned into initramfs-framework modules to reduce duplication.
@@ -33,6 +34,18 @@ early_setup() {
        cd -
     fi
 
+    if [ -d $MMC_DEVICE_PATH/mmcblk0p1/ ]; then
+        for part in `ls $MMC_DEVICE_PATH/ | grep mmcblk0p`
+        do
+            partname=`cat $MMC_DEVICE_PATH/$part/uevent | grep PARTNAME= | sed 's/.*=//'`
+            if [ "data" = "$partname" ] || [ "system" = "$partname" ] ||
+               [ "boot" = "$partname" ] || [ "recovery" = "$partname" ] ||
+               [ "misc" = "$partname" ]; then
+                ln -s /dev/$part  /dev/$partname
+            fi
+        done
+    fi
+
     mkdir -p /run
     mkdir -p /var/run
 
diff --git a/recipes-core/swupdate/aml-swupdate-ui/directfbrc b/recipes-core/swupdate/aml-swupdate-ui/ota_directfbrc
similarity index 78%
rename from recipes-core/swupdate/aml-swupdate-ui/directfbrc
rename to recipes-core/swupdate/aml-swupdate-ui/ota_directfbrc
index 1468c3b..b1ffd54 100755
--- a/recipes-core/swupdate/aml-swupdate-ui/directfbrc
+++ b/recipes-core/swupdate/aml-swupdate-ui/ota_directfbrc
@@ -2,13 +2,15 @@
 system        = fbdev
 fbdev         = /dev/fb0
 mode = 480x480
-depth = 16
-pixelformat = RGB16
+depth = 32
+pixelformat = ARGB
+no-sighandler
 #no-hardware
 hardware
 #accelerator=49
 # Disable Cursor
 no-cursor
+no-vt
 wm=default
 #debug
 #desktop-buffer-mode=backsystem
diff --git a/recipes-core/swupdate/aml-swupdate-ui_git.bb b/recipes-core/swupdate/aml-swupdate-ui_git.bb
index 84f8e80..723494e 100644
--- a/recipes-core/swupdate/aml-swupdate-ui_git.bb
+++ b/recipes-core/swupdate/aml-swupdate-ui_git.bb
@@ -1,7 +1,7 @@
 SUMMARY = "aml swupdate ui"
 LICENSE = "CLOSED"
 SRC_URI += "file://recovery.bmp \
-            file://directfbrc \
+            file://ota_directfbrc \
             file://recovery.jpg \
 "
 
@@ -16,11 +16,11 @@ S = "${WORKDIR}/git"
 
 # configurate lvgl app needs below two conditions
 # if you don't want to use lvgl, disable below two configurations.
-EXTRA_OEMAKE = "CONFIG_LVGL_APP=y"
-PACKAGECONFIG_append += " lvgl"
+#EXTRA_OEMAKE = "CONFIG_LVGL_APP=y"
+#PACKAGECONFIG_append += " lvgl"
 
 # if you use directfb, enable this configuration.
-# PACKAGECONFIG_append += " directfb"
+PACKAGECONFIG_append += " directfb"
 
 PACKAGECONFIG[lvgl] = "-lvgl,-no-lvgl,lvgl lv-drivers"
 PACKAGECONFIG[directfb] = "-directfb,-no-directfb,directfb"
@@ -41,7 +41,7 @@ do_install() {
     *)
         install -m 0644 ${WORKDIR}/recovery.bmp ${D}/etc
         install -d ${D}${sysconfdir}
-        install -m 0644 ${WORKDIR}/directfbrc ${D}/etc
+        install -m 0644 ${WORKDIR}/ota_directfbrc ${D}/etc
     ;;
     esac
 
diff --git a/recipes-core/swupdate/swupdate/swupdate.sh b/recipes-core/swupdate/swupdate/swupdate.sh
index 4f1b98f..0812b5f 100644
--- a/recipes-core/swupdate/swupdate/swupdate.sh
+++ b/recipes-core/swupdate/swupdate/swupdate.sh
@@ -23,16 +23,20 @@ case $system_mtd_number in
     *)
         ;;
 esac
-fbset -fb /dev/fb0 -g 480 480 480 960 16
-echo 0 0 479 479 > /sys/class/graphics/fb0/window_axis
-echo 0 0 479 479 > /sys/class/graphics/fb0/free_scale_axis
-echo 0x10001 > /sys/class/graphics/fb0/free_scale
-echo 0 > /sys/class/graphics/fb0/blank
-echo 1 > /sys/class/graphics/fb1/blank
-echo 1 > /sys/class/graphics/fb0/osd_clear
 
 swupdateui_run()
 {
+    # Self negotiating between drm and display, don't need to set here.
+    echo 0 > /sys/class/graphics/fb0/blank
+    echo 1 > /sys/class/graphics/fb1/blank
+
+    # Clean the default directfbrc file and add new configuration
+    # suitable for the upgrade.
+    if [ -f "/etc/directfbrc" ]; then
+        cat /dev/null > /etc/directfbrc
+        cat /etc/ota_directfbrc > /etc/directfbrc
+    fi
+
     if [ -f "/usr/bin/swupdateui" ]; then
         if [ -e "/etc/recovery.bmp" ]; then
             swupdateui /etc/recovery.bmp &
@@ -67,9 +71,7 @@ fi
 if [ -f "/mnt/software.swu" ]; then
     echo "find software.swu in data, now start update......"
     swupdateui_run
-    if [ -f "/proc/inand" ]; then
-        swupdate -l 6 -k /etc/swupdate-public.pem -i /mnt/software.swu
-    elif [ "${1}" = "ubifs" ]; then
+    if [ "${1}" = "ubifs" ]; then
         swupdate -l 6 -b "$str" -k /etc/swupdate-public.pem -i /mnt/software.swu
     else
         swupdate -l 6 -k /etc/swupdate-public.pem -i /mnt/software.swu
@@ -125,9 +127,7 @@ elif [ -f "$OTA_FILE_FLAG" ]; then
     #Here we detect if the OTA package exist or not
     if [ -n "$NETWORK_READY" ] && [ -n "$OTA_PACKAGE_READY" ]; then
         swupdateui_run
-        if [ -f "/proc/inand" ]; then
-            swupdate -l 6 -k /etc/swupdate-public.pem -D "-t 60"
-        elif [ "${1}" = "ubifs" ]; then
+        if [ "${1}" = "ubifs" ]; then
             swupdate -l 6 -b "$str" -k /etc/swupdate-public.pem -D "-t 60"
         else
             swupdate -l 6 -k /etc/swupdate-public.pem -D "-t 60"
@@ -188,9 +188,7 @@ else
     fi
     if [ -f "/run/media/$name/software.swu" ]; then
         swupdateui_run
-        if [ -f "/proc/inand" ]; then
-            swupdate -l 6 -k /etc/swupdate-public.pem -i /run/media/$name/software.swu
-        elif [ "${1}" = "ubifs" ]; then
+        if [ "${1}" = "ubifs" ]; then
             swupdate -l 6 -b "$str" -k /etc/swupdate-public.pem -i /run/media/$name/software.swu
         else
             swupdate -l 6 -k /etc/swupdate-public.pem -i /run/media/$name/software.swu
diff --git a/recipes-core/swupdate/swupdate_git.bb b/recipes-core/swupdate/swupdate_git.bb
index 7c506da..bac5035 100644
--- a/recipes-core/swupdate/swupdate_git.bb
+++ b/recipes-core/swupdate/swupdate_git.bb
@@ -236,11 +236,12 @@ do_install () {
     then
         install -m 0644 ${WORKDIR}/swupdate_nand.service ${D}/${systemd_unitdir}/system/swupdate.service
         sed 's@rootfs_type@${ROOTFS_TYPE}@' -i ${D}/${systemd_unitdir}/system/swupdate.service
-    else 
+    else
         install -m 0644 ${WORKDIR}/swupdate_emmc.service ${D}/${systemd_unitdir}/system/swupdate.service
     fi
     # config boardname
     sed 's@boardname@${MACHINE_ARCH}@' -i ${D}/etc/hwrevision
+    sed 's/_5.15//g' -i ${D}/etc/hwrevision
     sed 's@_lib32_@_@' -i ${D}/etc/hwrevision
 }
 
-- 
2.25.1

